File: sdks/local-rp/java/src/test/java/community/catalyst/linkkeys/localrp/KeysConformanceTest.java
package community.catalyst.linkkeys.localrp;

import static org.junit.jupiter.api.Assertions.assertArrayEquals;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.nio.charset.StandardCharsets;

import org.junit.jupiter.api.Test;

import community.catalyst.linkkeys.localrp.crypto.Crypto;
import community.catalyst.linkkeys.localrp.testutil.Fixtures;
import community.catalyst.linkkeys.localrp.testutil.MiniJson.JsonValue;

/** Conformance vectors: {@code keys.json}. */
class KeysConformanceTest {

    @Test
    void fingerprintsRoundTripThroughSdkFingerprintHelpers() {
        JsonValue d = Fixtures.load("keys.json");

        for (JsonValue node : new JsonValue[] {d.get("local_rp").get("signing"), d.get("domain_signing_key")}) {
            byte[] seed = Fixtures.hex(node.get("seed_hex").asString());
            byte[] expectedPublic = Fixtures.hex(node.get("public_key_hex").asString());
            String expectedFp = node.get("fingerprint_hex").asString();

            // Confirm seed and public key correspond to the same Ed25519
            // keypair: signing with the seed and verifying against the
            // fixture's public key must succeed.
            byte[] message = "conformance-check".getBytes(StandardCharsets.UTF_8);
            byte[] sig = Crypto.signEd25519(message, seed);
            assertTrue(Crypto.verifyEd25519(message, sig, expectedPublic));

            String computed = Crypto.fingerprint(expectedPublic);
            assertEquals(expectedFp, computed);

            // Round-trip through the SDK's own fingerprint string helpers.
            String s = Identity.fingerprintToString(computed);
            assertEquals(expectedFp, Identity.fingerprintFromString(s));

            assertEquals(32, seed.length);
        }

        // fingerprintFromString must reject non-fingerprint strings even
        // when they happen to be valid hex of the wrong length.
        assertThrows(SdkException.class, () -> Identity.fingerprintFromString("deadbeef"));
    }

    @Test
    void x25519PublicKeysDeriveFromPrivateKeys() {
        JsonValue d = Fixtures.load("keys.json");
        for (JsonValue node : new JsonValue[] {d.get("local_rp").get("encryption"), d.get("domain_encryption_recipient")}) {
            byte[] priv = Fixtures.hex(node.get("private_key_hex").asString());
            byte[] expectedPublic = Fixtures.hex(node.get("public_key_hex").asString());
            assertArrayEquals(expectedPublic, Crypto.derivePublicFromX25519Private(priv));
        }
    }
}

File: sdks/local-rp/kotlin/src/test/kotlin/community/catalyst/linkkeys/localrp/kt/KeysConformanceTest.kt
package community.catalyst.linkkeys.localrp.kt

import java.nio.charset.StandardCharsets
import org.junit.jupiter.api.Assertions.assertArrayEquals
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertThrows
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test
import community.catalyst.linkkeys.localrp.kt.protocol.Crypto
import community.catalyst.linkkeys.localrp.kt.testutil.Fixtures

/** Conformance vectors: `keys.json`, exercised through this package's own [Crypto] and [LocalRpIdentity] fingerprint helpers. */
class KeysConformanceTest {

    @Test
    fun fingerprintsRoundTripThroughSdkFingerprintHelpers() {
        val d = Fixtures.load("keys.json")

        for (node in listOf(d.get("local_rp").get("signing"), d.get("domain_signing_key"))) {
            val seed = Fixtures.hex(node.get("seed_hex").asString())
            val expectedPublic = Fixtures.hex(node.get("public_key_hex").asString())
            val expectedFp = node.get("fingerprint_hex").asString()

            // Confirm seed and public key correspond to the same Ed25519
            // keypair: signing with the seed and verifying against the
            // fixture's public key must succeed.
            val message = "conformance-check".toByteArray(StandardCharsets.UTF_8)
            val sig = Crypto.signEd25519(message, seed)
            assertTrue(Crypto.verifyEd25519(message, sig, expectedPublic))

            val computed = Crypto.fingerprint(expectedPublic)
            assertEquals(expectedFp, computed)

            // Round-trip through the SDK's own fingerprint string helpers.
            val s = fingerprintToString(computed)
            assertEquals(expectedFp, fingerprintFromString(s))

            assertEquals(32, seed.size)
        }

        // fingerprintFromString must reject non-fingerprint strings even
        // when they happen to be valid hex of the wrong length.
        assertThrows(LocalRpException.InvalidInput::class.java) { fingerprintFromString("deadbeef") }
    }

    @Test
    fun x25519PublicKeysDeriveFromPrivateKeys() {
        val d = Fixtures.load("keys.json")
        for (node in listOf(d.get("local_rp").get("encryption"), d.get("domain_encryption_recipient"))) {
            val priv = Fixtures.hex(node.get("private_key_hex").asString())
            val expectedPublic = Fixtures.hex(node.get("public_key_hex").asString())
            assertArrayEquals(expectedPublic, Crypto.derivePublicFromX25519Private(priv))
        }
    }
}
